File: 437/path-sum-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def __init__(self, count):
        self.count = count

    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: int
        """
        if root is None:
            return 0

        self.count = 0

        def dfs(node, tmp):
            if tmp == sum:
                self.count += 1
            if node.left is not None:
                return dfs(node.left, tmp + node.left.val)
            if node.right is not None:
                return dfs(node.left, tmp + node.right.val)

        dfs(root, root.val)
        if root.left is not None:
            self.count += self.pathSum(root.left, sum)
        if root.right is not None:
            self.count += self.pathSum(root.right, sum)

        return self.count

File: 437/path-sum-iii.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number}
 */
function pathSum(root, sum) {
    if (!root)
        return 0;

    let ans = 0;

    function dfs(node, tmp) {
        tmp === sum && (ans++);
        node.left && dfs(node.left, tmp + node.left.val);
        node.right && dfs(node.right, tmp + node.right.val);
    }

    dfs(root, root.val);
    root.left && (ans += pathSum(root.left, sum));
    root.right && (ans += pathSum(root.right, sum));

    return ans;
}